infra/Pos/Util/LogSafe.hs

```diff
@@ -52,13 +111,207 @@ logNoticeS  = logMessageS Notice
 logWarningS = logMessageS Warning
 logErrorS   = logMessageS Error
 
--- | Same as 'logMesssage', but log to two loggers, put only insecure
--- version to memmode.
+-- | Same as 'logMesssage', but log to secret logs, put only insecure
+-- version to memmode (to terminal).
 logMessageS
     :: (HasLoggerName m, MonadIO m)
     => Severity
     -> Text
     -> m ()
-logMessageS severity t = execSecureLogWrapped $ do
-    name <- getLoggerName
-    dispatchMessage name severity t
+logMessageS severity t =
+    reify selectSecretLogs $ \s ->
+    execSecureLogWrapped s $ do
+        name <- getLoggerName
+        dispatchMessage name severity t
+
+----------------------------------------------------------------------------
+-- Secure buildables
+----------------------------------------------------------------------------
+
+-- * Secutity log levels
+
+-- Stuff below provides way to use different 'instance Buildable' for writing
+-- to secret and public logs.
+--
+-- Example:
+--
+-- @
+-- logInfoSP $ \sl -> ("Nyan: "%secureF sl build) Nyan
+-- @
+--
+-- this function prints value to secret logs as is, while for public logs it uses
+-- 'instance Buildable (PublicLog Nyan)'.
+--
+-- By analogy, 'logInfoP' will log to public logs only, requiring
+-- 'instance Buildable (PublicLog ...)'. It also provides security level @sl@
+-- parameter which is redundant, but allows to quickly switch
+-- between 'logInfoSP' and 'logInfoP' and protects you from forgetting to use
+-- secure 'Buildable'.
+
+-- | This security level is used when we known a priori that the text is builded
+-- for secret logs, and we can use exact `instance Buildable`.
+data SecLogLevel
+
+-- | Similar to 'SecLogLevel', requires special secure `instance Buildable`.
+data PubLogLevel
+
+-- | Needed to define "closed type instance" below.
+-- It's used when exact logging security level it to be defined later,
+-- and we have to "be ready" to provide ways to build text for both secret and
+-- public logs.
+data AltLogLevel s where
+    AltSecLogLevel :: AltLogLevel SecLogLevel
+    AltPubLogLevel :: AltLogLevel PubLogLevel
+
+-- | Promotes 'AltLogLevel' constructors to type.
+data ReflectedAltLogLevel rs
+
+
+-- Modifies 'instance Buildable' to correspond given security level @ s @.
+data SecureLog s a = SecureLog
+    { getSecureLog :: a
+    } deriving (Eq, Ord)
+
+secureLog :: Proxy s -> a -> SecureLog s a
+secureLog _ = SecureLog
+
+coerce :: forall s2 s1 a. SecureLog s1 a -> SecureLog s2 a
+coerce = SecureLog . getSecureLog
```

Comment:
Give it a different name, something like `coerceSecureLog`, because when I see `coerce` I expect it to be the one from `Data.Coerce`. Actually, @ChShersh, would you please add `coerce` from `Data.Coerce` to Universum?